win32/src/async_socket_win32.c

```diff
@@ -645,153 +646,144 @@ int async_socket_receive_async(ASYNC_SOCKET_HANDLE async_socket, ASYNC_SOCKET_BU
+                (payload[i].length == 0)
+                )
+            {
+                LogError("Invalid buffer %" PRIu32 ": buffer=%p, length = %" PRIu32, i, payload[i].buffer, payload[i].length);
+                break;
+            }
+
+            if (total_buffer_bytes + payload[i].length < total_buffer_bytes)
+            {
+                /* Codes_SRS_ASYNC_SOCKET_WIN32_01_096: [ If the sum of buffer lengths for all the buffers in payload is greater than UINT32_MAX, async_socket_receive_async shall fail and return a non-zero value. ]*/
+                LogError("Overflow in total buffer length computation");
+                break;
+            }
+            else
+            {
+                total_buffer_bytes += payload[i].length;
+            }
+        }
+
+        if (i < buffer_count)
+        {
+            LogError("Invalid buffers passed to async_socket_receive_async");
             result = MU_FAILURE;
         }
         else
         {
-            uint32_t i;
-            uint32_t total_buffer_bytes = 0;
+            (void)InterlockedIncrement(&async_socket->pending_api_calls);
 
-            for (i = 0; i < buffer_count; i++)
+            if (InterlockedAdd(&async_socket->state, 0) != (LONG)ASYNC_SOCKET_WIN32_STATE_OPEN)
             {
-                if (
-                    /* Codes_SRS_ASYNC_SOCKET_WIN32_01_091: [ If any of the buffers in payload has buffer set to NULL, async_socket_receive_async shall fail and return a non-zero value. ]*/
-                    (payload[i].buffer == NULL) ||
-                    /* Codes_SRS_ASYNC_SOCKET_WIN32_01_092: [ If any of the buffers in payload has length set to 0, async_socket_receive_async shall fail and return a non-zero value. ]*/
-                    (payload[i].length == 0)
-                    )
-                {
-                    LogError("Invalid buffer %" PRIu32 ": buffer=%p, length = %" PRIu32, i, payload[i].buffer, payload[i].length);
-                    break;
-                }
-
-                if (total_buffer_bytes + payload[i].length < total_buffer_bytes)
-                {
-                    /* Codes_SRS_ASYNC_SOCKET_WIN32_01_096: [ If the sum of buffer lengths for all the buffers in payload is greater than UINT32_MAX, async_socket_receive_async shall fail and return a non-zero value. ]*/
-                    LogError("Overflow in total buffer length computation");
-                    break;
-                }
-                else
-                {
-                    total_buffer_bytes += payload[i].length;
-                }
-            }
-
-            if (i < buffer_count)
-            {
-                LogError("Invalid buffers passed to async_socket_receive_async");
+                /* Codes_SRS_ASYNC_SOCKET_WIN32_01_098: [ If async_socket is not OPEN, async_socket_receive_async shall fail and return a non-zero value. ]*/
+                LogWarning("Not open");

```

Comment:
>"Not open" [](http://example.com/codeflow?start=27&length=10)

It would be really nice to add the current state in here so it can give a clue as to where we are when we failed.

Reply:
Change here and everywhere where "not open" could be found.